File: app/src/test/java/io/intrepid/contest/screens/entrysubmission/entryname/EntryNamePresenterTest.java
package io.intrepid.contest.screens.entrysubmission.entryname;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.junit.MockitoJUnitRunner;

import io.intrepid.contest.screens.entrysubmission.entryname.EntryNameContract.Presenter;
import io.intrepid.contest.screens.entrysubmission.entryname.EntryNameContract.View;
import io.intrepid.contest.testutils.BasePresenterTest;
import io.intrepid.contest.testutils.TestPresenterConfiguration;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class EntryNamePresenterTest extends BasePresenterTest<EntryNamePresenter> {
    @Mock
    View mockView;

    private Presenter presenter;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        presenter = new EntryNamePresenter(mockView, TestPresenterConfiguration.createTestConfiguration());
    }

    @Test
    public void onViewCreatedShouldShowWelcomeMessage() {
        presenter.onViewCreated();
        verify(mockView).showWelcomeMessage();
    }

    @Test
    public void onEntryNameTextViewFocusChangedShouldShowNextButtonWhenTextViewIsFocused() {
        presenter.onEntryNameFocusChanged(true);

        verify(mockView).showEntryNameButton();
        verify(mockView, never()).hideEntryNameButton();
    }

    @Test
    public void onEntryNameTextViewFocusChangedShouldHideNextButtonWhenTextViewIsNotFocused() {
        presenter.onEntryNameFocusChanged(false);

        verify(mockView).hideEntryNameButton();
        verify(mockView, never()).showEntryNameButton();
    }

    @Test
    public void onEntryNameTextViewChangedShouldDisableNextButtonWhenNameIsNotEmpty() {
        String newText = "";

        presenter.onEntryNameTextChanged(newText);

        verify(mockView).disableEntryNameButton();
        verify(mockView, never()).enableEntryNameButton();
    }

    @Test
    public void onEntryNameTextViewChangedShouldEnableNextButtonWhenNameIsNotEmpty() {
        String newText = "1";

        presenter.onEntryNameTextChanged(newText);

        verify(mockView).enableEntryNameButton();
        verify(mockView, never()).disableEntryNameButton();
    }

    @Test
    public void onEntryNameSubmittedShouldShowEntryImageScreen() {
        String entryName = "Name";
        presenter.onEntryNameSubmitted(entryName);
        verify(mockView).showEntryImageScreen(any());
    }
}

File: app/src/test/java/io/intrepid/contest/screens/entrysubmission/entryname/EntryNamePresenterTest.kt
package io.intrepid.contest.screens.entrysubmission.entryname

import io.intrepid.contest.screens.entrysubmission.entryname.EntryNameContract.View
import io.intrepid.contest.testutils.BasePresenterTest
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class EntryNamePresenterTest : BasePresenterTest<EntryNamePresenter>() {
    @Mock
    private lateinit var mockView: View

    @Before
    fun setup() {
        presenter = EntryNamePresenter(mockView, testConfiguration)
    }

    @Test
    fun onEntryNameTextViewChangedShouldHideNextButtonWhenNameIsNotEmpty() {
        val newText = ""

        presenter.onEntryNameTextChanged(newText)

        verify<View>(mockView).hideEntryNameButton()
        verify<View>(mockView, never()).showEntryNameButton()
    }

    @Test
    fun onEntryNameTextViewChangedShouldShowIconWhenNameIsNotEmpty() {
        val newText = ""

        presenter.onEntryNameTextChanged(newText)

        verify<View>(mockView).showEntryNameIcon()
        verify<View>(mockView, never()).hideEntryNameIcon()
    }

    @Test
    fun onEntryNameTextViewChangedShouldShowNextButtonWhenNameIsNotEmpty() {
        val newText = "1"

        presenter.onEntryNameTextChanged(newText)

        verify<View>(mockView).showEntryNameButton()
        verify<View>(mockView, never()).hideEntryNameButton()
    }

    @Test
    fun onEntryNameTextViewChangedShouldHideIconWhenNameIsNotEmpty() {
        val newText = "1"

        presenter.onEntryNameTextChanged(newText)

        verify<View>(mockView).hideEntryNameIcon()
        verify<View>(mockView, never()).showEntryNameIcon()
    }

    @Test
    fun onEntryNameSubmittedShouldShowEntryImageScreen() {
        val entryName = "Name"
        presenter.onEntryNameSubmitted(entryName)
        verify<View>(mockView).showEntryImageScreen(any<String>())
    }
}
